code/game/objects/structures/crates_lockers/closets/coffin.dm

```diff
@@ -6,13 +6,143 @@
+	if (src.opened && is_locking(mob_lock_type)) //only unbuckle if you are buckled in in the first place
+		manual_unbuckle(usr)
+		setDensity(FALSE) //this is needed for some reason 
+		return
+	var/mob/closet_dweller = locate() in src.loc
+	if (src.opened && closet_dweller) //buckle only the mob inside the closet
+		buckle_mob(closet_dweller, usr)
+
+/obj/structure/closet/coffin/proc/handle_user_visibility() //after each open/close action assert the correct user visibility
+	if (!get_locked(mob_lock_type).len)
+		return
+	var/mob/locked = get_locked(mob_lock_type)[1]	
+	if (src.opened)  
+		locked.alphas["coffin_invis"] = 255
+		locked.handle_alpha()
+	else 
+		locked.alphas["coffin_invis"] = 1
+		locked.handle_alpha()
+
+/obj/structure/closet/coffin/relaymove(mob/user as mob)
+	if (has_locked_mobs())
+		return
+	..()
+
+/obj/structure/closet/coffin/proc/manual_unbuckle(var/mob/user)
+	if(user.size <= SIZE_TINY)
+		to_chat(user, "<span class='warning'>You are too small to do that.</span>")
+		return FALSE
+
+	if(is_locking(mob_lock_type))
+		add_fingerprint(user)
+
+		var/mob/M = get_locked(mob_lock_type)[1]
+		var/success = unlock_atom(M)
+
+		if(M != user)
+			if(!success)
+				user.delayNextAttack(8)
+				M.visible_message("<span class='warning'>[user] struggles in vain trying to pull [M] off \the [src].</span>")
+				return FALSE
+			M.visible_message(
+				"<span class='notice'>[M] was unbuckled by [user]!</span>",
+				"You were unbuckled from \the [src] by [user].",
+				"You hear metal clanking.")
+		else
+			if(!success)
+				user.delayNextAttack(8)
+				M.visible_message("<span class='warning'>[user] struggles in vain trying to pull themselves off \the [src].</span>")
+				return FALSE
+			M.visible_message(
+				"<span class='notice'>[M] unbuckled \himself!</span>",
+				"You unbuckle yourself from \the [src].",
+				"You hear metal clanking.")
+		playsound(src, 'sound/misc/buckle_unclick.ogg', 50, 1)
+		return TRUE
+
+/obj/structure/closet/coffin/proc/buckle_mob(mob/M as mob, mob/user as mob)
+	if(!Adjacent(user) || user.incapacitated() || istype(user, /mob/living/silicon/pai))
+		return
+
+	if(!ismob(M) || (M.loc != src.loc)  || M.locked_to)
+		return
+
+	for(var/mob/living/L in get_locked(mob_lock_type))
+		to_chat(user, "<span class='warning'>Somebody else is already buckled into \the [src]!</span>")
+		return
+
+	if(user.size <= SIZE_TINY) //Fuck off mice
+		to_chat(user, "<span class='warning'>You are too small to do that.</span>")
+		return
+
+	if(isanimal(M))
+		if(M.size <= SIZE_TINY) //Fuck off mice
+			to_chat(user, "<span class='warning'>The [M] is too small to buckle in.</span>")
+			return
+
+	if(istype(M, /mob/living/carbon/slime))
+		to_chat(user, "<span class='warning'>The [M] is too squishy to buckle in.</span>")
+		return
+
+	if(M == usr)
+		M.visible_message(\
+			"<span class='notice'>[M.name] buckles in!</span>",\
```

Comment:
Rather than using `[M.name]`, prefer `\The [M]` or `\the [M]`. 
Putting a mob into brackets is more or less the equivalent of a `ToString()` call in modern languages.